### DISCUSSION TITLE
TypeScript - What is the correct way to type custom model methods & item methods?

### DISCUSSION THREAD

**Discussion by tranhl** [dynamoose/dynamoose#1646] [Q&A] [answered]
As per title. Let's say I have a model called `Foo`, and I attach the following custom methods:

```ts
const Foo = model<FooItem>()

Foo.methods.set('bar', function () {})
Foo.methods.item.set('baz', function() {})
```

What is the correct way to type `bar` and `baz`?
As a workaround, right now I'm passing the type of `bar` into the `model` function's generics:

```ts
const Foo = model<FooItem & { bar: () => void }>()

Foo.methods.set('bar', function () {})
Foo.methods.item.set('baz', function() {})
```

However, it doesn't seem possible to type `baz` using the same method, as the generic is the type for the model, not the item instance. The model and items would share the same custom methods, which wouldn't be correct.

What would be the correct way to approach this?

**Comment by fishcharlie** [comment, marked as answer]
@tranhl I don't think this case is supported currently. You either have to type cast it to `any` or use `ts-ignore`. Dynamoose TypeScript support is currently in beta: https://dynamoosejs.com/getting_started/TypeScript.

PRs would be appreciated here. Or an issue documenting this feature request.
